extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -77,20 +85,53 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     @Override
     public void initialize() {
         TradfriGatewayConfig configuration = getConfigAs(TradfriGatewayConfig.class);
-        if (configuration.host == null || configuration.host.isEmpty()) {
+
+        if (isNullOrEmpty(configuration.host)) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                     "Host must be specified in the configuration!");
             return;
         }
-        if (configuration.code == null || configuration.code.isEmpty()) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "Security code must be provided in the configuration!");
-            return;
+        if (isNullOrEmpty(configuration.code)) {
+            if (isNullOrEmpty(configuration.identity) || isNullOrEmpty(configuration.preSharedKey)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Either security code or identity and pre-shared key must be provided in the configuration!");
+                return;
+            }
+        } else {
+            if (isOldFirmware()) {
+                /*
+                 * older firmware - fall back to authentication with security code
+                 * in this case the Thing configuration will not be persisted
+                 */
+                logger.warn("Gateway with old firmware - please consider upgrading to the latest version.");
+                configuration.identity = "";
+                configuration.preSharedKey = configuration.code;
+            } else {
+                configuration.identity = UUID.randomUUID().toString().replace("-", "");
+                configuration.preSharedKey = authenticateOnGateway(configuration);
```

Comment:
What happened to all our discussion that this is a potentially long running operation that must not be done here?

Reply:
you said: 
>Isn't that a bit too much at once?

Reply:
I meant that you must not move ALL the code into the async job...

Reply:
Sorry but I can't imagine how to run an async task, retrieve the result and continue the initialization.

If you expect to finish the initialize method before all the procedures of authenticating on the gateway and retrieving a psk are finished - I don't think this is quite good. After all - status: ```CONFIGURATION_PENDING``` is a lot more speaking than status ```UNKNOWN```.

And leaving only the NULL checks in the initialize and then calling another async method is I think - meaningless - anyway all stuff have to be complete before the thing can go operational - so it will be async only from framework point of view.

Reply:
Maybe you should have read my full comment above:

> Isn't that a bit too much at once? **The whole config checking should be done synchronously. If something is wrong there (or everything is alright), there is no need to schedule any additional job at all.**